packages/beacon-node/src/network/reqresp/beaconBlocksMaybeBlobsByRoot.ts

```diff
@@ -39,3 +39,50 @@ export async function beaconBlocksMaybeBlobsByRoot(
   // and here it should be infinity since all bobs should match
   return matchBlockWithBlobs(config, allBlocks, allBlobSidecars, Infinity, BlockSource.byRoot);
 }
+
+export async function unavailableBeaconBlobsByRoot(
+  config: ChainForkConfig,
+  network: INetwork,
+  peerId: PeerIdStr,
+  unavailableBlockInput: BlockInput
+): Promise<BlockInput> {
+  if (unavailableBlockInput.type !== BlockInputType.blobsPromise) {
+    return unavailableBlockInput;
+  }
+
+  const blobIdentifiers: deneb.BlobIdentifier[] = [];
+  const {block, blobsCache, resolveAvailability, blockBytes} = unavailableBlockInput;
+
+  const slot = block.message.slot;
+  const blockRoot = config.getForkTypes(slot).BeaconBlock.hashTreeRoot(block.message);
+
+  const blobKzgCommitmentsLen = (block.message.body as deneb.BeaconBlockBody).blobKzgCommitments.length;
+  for (let index = 0; index < blobKzgCommitmentsLen; index++) {
+    if (blobsCache.has(index) === false) blobIdentifiers.push({blockRoot, index});
+  }
+
+  let allBlobSidecars: deneb.BlobSidecar[];
+  if (blobIdentifiers.length > 0) {
+    allBlobSidecars = await network.sendBlobSidecarsByRoot(peerId, blobIdentifiers);
+  } else {
+    allBlobSidecars = [];
+  }
+
+  // add them in cache so that its reflected in all the blockInputs that carry this
+  // for e.g. a blockInput that might be awaiting blobs promise fullfillment in
+  // verifyBlocksDataAvailability
+  for (const blobSidecar of allBlobSidecars) {
+    blobsCache.set(blobSidecar.index, {blobSidecar, blobBytes: null});
+  }
+
+  // check and see if all blobs are now available and in that case resolve availability
+  // if not this will error and the leftover blobs will be tried from another peer
+  const allBlobs = getBlockInputBlobs(blobsCache);
+  const {blobs, blobsBytes} = allBlobs;
+  if (blobs.length !== blobKzgCommitmentsLen) {
+    throw Error(`Not all blobs fetched missingBlobs=${blobKzgCommitmentsLen - blobs.length}`);
+  }
+
+  resolveAvailability(allBlobs);
```

Comment:
track the source of this `resolveAvailability()` call, is it gossip or this on demand call. I think this is the most important metric coming out of this work.

Reply:
adding this